### PR TITLE
Remove extra cast to PSObject in ToStringEnumerable() method

### LINKED ISSUE #10377: Performance: -join slower than [string]::Join()
# Steps to reproduce

Joining 100kb random numbers, the operator is much slower, ~13x here:

```powershell
[string[]]$nums = Get-Random -Minimum 1 -Maximum 100 -Count 100kb

measure-command { $nums -join ','            } |% TotalMilliseconds
measure-command { [string]::join(',', $nums) } |% TotalMilliseconds

234.82690000000002
14.497100000000001
```

Stranger, with [a big wordlist text file](https://github.com/dolph/dictionary/blob/master/enable1.txt?raw=true), loading it two different ways seems to make a difference:

```powershell
$w1 = Get-Content d:\test\wordlist.txt;   $w1.Count
$w2 = ${D:\test\wordlist.txt};            $w2.Count

measure-command { $w1 -join "`r`n" } |% TotalMilliseconds
measure-command { $w2 -join "`r`n" } |% TotalMilliseconds

measure-command { [string]::join("`r`n", $w1) } |% TotalMilliseconds
measure-command { [string]::join("`r`n", $w2) } |% TotalMilliseconds

172823
172823        # same word count

254.6534
423.63100000000003    # <-- slower?

183.01510000000002
16.148400000000002    # <-- how come this one is SO fast?
```

# Environment data

```none
PS D:\> $PSVersionTable

Name                           Value
----                           -----
PSVersion                      7.0.0-preview.1
PSEdition                      Core
GitCommitId                    7.0.0-preview.1
OS                             Microsoft Windows 10.0.17763
Platform                       Win32NT
PSCompatibleVersions           {1.0, 2.0, 3.0, 4.0…}
PSRemotingProtocolVersion      2.3
SerializationVersion           1.1.0.1
WSManStackVersion              3.0
```

### PR DESCRIPTION
**Pull request by iSazonov**
# PR Summary

- Remove extra cast to PSObject in ToStringEnumerable() method
- Remove extra Remove() call

PerfView screenshot: left - before the change, right - after the fix.
image

## PR Context

Fix #10377

For the scenario:
```powershell
$w2=${C:\tmp\source2.csv}
for ($i=1; $i -le 100; $i++) { $tmp = $w2 -join "`r`n" }
```
in `$w2` we have an array of strings (_not PSObjects!_) and join the strings. Before the fix in ToStringEnumerable() we did extra cast every string from $w2 to PSObject which caused unnecessary memory allocations and slow down the scenario because of skipping fast code path for value types.

## PR Checklist

- [x] [PR has a meaningful title](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
    - Use the present tense and imperative mood when describing your changes
- [x] [Summarized changes](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
- [x] [Make sure all `.h`, `.cpp`, `.cs`, `.ps1` and `.psm1` files have the correct copyright header](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
- [x] This PR is ready to merge and is not [Work in Progress](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---work-in-progress).
    - If the PR is work in progress, please add the prefix `WIP:` or `[ WIP ]` to the beginning of the title (the `WIP` bot will keep its status check at `Pending` while the prefix is present) and remove the prefix when the PR is ready.
- **[Breaking changes](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#making-breaking-changes)**
    - [x] None
    - **OR**
    - [ ] [Experimental feature(s) needed](https://github.com/MicrosoftDocs/PowerShell-Docs/blob/staging/reference/6/Microsoft.PowerShell.Core/About/about_Experimental_Features.md)
        - [ ] Experimental feature name(s): 
- **User-facing changes**
    - [x] Not Applicable
    - **OR**
    - [ ] [Documentation needed](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
        - [ ] Issue filed: 
- **Testing - New and feature**
    - [ ] N/A or can only be tested interactively
    - **OR**
    - [x] [Make sure you've added a new test if existing tests do not effectively test the code changed](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#before-submitting)
- **Tooling**
    - [ ] I have considered the user experience from a tooling perspective and don't believe tooling will be impacted.
    - **OR**
    - [ ] I have considered the user experience from a tooling perspective and enumerated concerns in the summary. This may include:
        - Impact on [PowerShell Editor Services](https://github.com/PowerShell/PowerShellEditorServices) which is used in the [PowerShell extension](https://github.com/PowerShell/vscode-powershell) for VSCode (which runs in a different PS Host).
        - Impact on Completions (both in the console and in editors) - one of PowerShell's most powerful features.
        - Impact on [PSScriptAnalyzer](https://github.com/PowerShell/PSScriptAnalyzer) (which provides linting & formatting in the editor extensions).
        - Impact on [EditorSyntax](https://github.com/PowerShell/EditorSyntax) (which provides syntax highlighting with in VSCode, GitHub, and many other editors).

### CODE REVIEW

File: src/System.Management.Automation/engine/MshObject.cs [hunk 1]
```diff
@@ -1144,8 +1144,7 @@ internal static string ToStringEnumerable(ExecutionContext context, IEnumerable
             {
                 if (obj != null)
                 {
-                    PSObject mshObj = PSObject.AsPSObject(obj);
-                    returnValue.Append(PSObject.ToString(context, mshObj, separator, format, formatProvider, false, false));
+                    returnValue.Append(PSObject.ToString(context, obj, separator, format, formatProvider, false, false));
```
Comment (daxian-dbw) [inline src/System.Management.Automation/engine/MshObject.cs]: By wrapping the object into a `PSObject`, `PSObject.ToString` will be able to find the `ToString(...)` script method/code method defined for the type of the object in `types.ps1xml` files, so as to use that in precedence.
So simply removing this step will be a breaking change.

I looked at this issue myself, and currently `ToStringEnumerator` and `ToStringEnumerable` behave differently in terms of how the `ToString(...)` methods defined in `types.ps1xml` files are honored for the element object.
From the completeness perspective, `ToStringEnumerator` should do the same as `ToStringEnumerable` -- wrapping the object before passing to `PSObject.ToString`.

A simpler and narrow-scoped fix would be to special case string objects:
```c#
if (obj != null)
{
    object mshObj = obj -is string ? obj : PSObject.AsPSObject(obj);
    returnValue.Append(PSObject.ToString(context, mshObj, separator, format, formatProvider, false, false));
}
```

**Review by daxian-dbw — CHANGES_REQUESTED** [review]

File: src/System.Management.Automation/engine/MshObject.cs [hunk 2]
```diff
@@ -1156,9 +1155,7 @@ internal static string ToStringEnumerable(ExecutionContext context, IEnumerable
                 return string.Empty;
             }
 
-            int separatorLength = separatorToUse.Length;
-            returnValue.Remove(returnValue.Length - separatorLength, separatorLength);
-            return returnValue.ToString();
+            return returnValue.ToString(0, returnValue.Length - separatorToUse.Length);
```
Comment (daxian-dbw) [inline src/System.Management.Automation/engine/MshObject.cs]: You may want to change the similar code in `ToStringEnumerator` as well.

**Comment by daxian-dbw** [comment]
> Before the fix in ToStringEnumerable() we did extra cast every string from $w2 to PSObject which caused unnecessary memory allocations and slow down the scenario because of skipping fast code path for value types.

You will need to update this PR description since casting to `PSObject` is intentional. Also, please list the execution time before/after your changes.
It would be great if you can add a test to cover the case where a `ToString` method is defined in a type `.ps1xml` file for a type, and see if `-join` honors that method.

File: src/System.Management.Automation/engine/MshObject.cs (hunk 1 above)
Comment (iSazonov) [reply src/System.Management.Automation/engine/MshObject.cs]: My thoughts were that:
- having TryFastTrackPrimitiveTypes() implies that we don't want to find ToString() in types.ps1xml for _primitive_ types.
- it is unlikely breaking change. Why would anybody want to define custom ToString() for primitive types? I see no use. Even if so then I believe right way is to define a custom type with custom ToString().
- only `-Join` is affected and affected only on half as you mentioned - ToStringEnumerator() does not cast to PSObject. So we should fix either ToStringEnumerable() or ToStringEnumerator().

I tend to accept the unlikely breaking change and fix ToStringEnumerable(), not  ToStringEnumerator() - we get more benefits (great perf win as minimum).

**Comment by iSazonov** [comment]
I will add perf results after we get a consensus about that the fix should be.

File: src/System.Management.Automation/engine/MshObject.cs (hunk 2 above)
Comment (iSazonov) [reply src/System.Management.Automation/engine/MshObject.cs]: Done.

File: src/System.Management.Automation/engine/MshObject.cs (hunk 1 above)
Comment (daxian-dbw) [reply src/System.Management.Automation/engine/MshObject.cs]: > Why would anybody want to define custom ToString() for primitive types? 

The thing is, you don't know the element of `enumerable` here is a primitive type. I can perfectly do `(Get-ChildItem c:\) -join ";"`.
For primitive types, I agree it's unlikely that a `ToString` method is defined in ETS for them, but they are more likely than `string` (yes, `string` is not a primitive type).

> only -Join is affected and affected only on half as you mentioned

This is not true. `PSObject.ToStringEnumerable` is used in `PSObject.ToString` recursively, and thus any calls to `PSObject.ToString` will be affected.

> ToStringEnumerator() does not cast to PSObject. So we should fix either ToStringEnumerable() or ToStringEnumerator().

`ToStringEnumerator()` is obviously the one that should be fixed. It's missing a functionality that powershell should have provided -- honor the custom `ToString` method defined in ETS for a type.

Comment (iSazonov) [reply src/System.Management.Automation/engine/MshObject.cs]: Do you mean that the PR broke `(Get-ChildItem c:\) -join ";"` ? But this works as expected. Could you please show an example which the PR broke?

Comment (iSazonov) [reply src/System.Management.Automation/engine/MshObject.cs]: Thanks for explanations!
If we call PSObject.ToString() for non-PSObject object we always falls in IFormattable code path as last resort and don't honor ETS - is it right?

The IFormattable code path doesn't allow me to fix PSObject.ToString() in universal way so I think we could call TryFastTrackPrimitiveTypes() in ToStringEnumerable() (and ToStringEnumerator()) before casting to PSObject and  calling PSObject.ToString().

File: src/System.Management.Automation/engine/MshObject.cs [hunk 3]
```diff
@@ -1122,18 +1122,24 @@ internal static string ToStringEnumerator(ExecutionContext context, IEnumerator
...
+                if (TryFastTrackPrimitiveTypes(obj, format, formatProvider, out string objString))
+                {
+                    returnValue.Append(objString);
+                }
+                else
+                {
+                    PSObject mshObj = PSObject.AsPSObject(obj);
+                    returnValue.Append(PSObject.ToString(context, mshObj, separator, format, formatProvider, false, false));
+                    returnValue.Append(separatorToUse);
+                }
```
Comment (daxian-dbw) [inline src/System.Management.Automation/engine/MshObject.cs]: `DateTime` has a higher chance to get a custom `ToString` definition in ETS.
I think we should just strictly restrict the special handling to `string` and true primitive types (namely `type.IsPrimitive` is true).
```c#
var objType = obj.GetType();
if (objType != typeof(string) && !objType.IsPrimitive)
{
    obj = PSObject.AsPSObject(obj);
}

returnValue.Append(PSObject.ToString(context, obj, separator, format, formatProvider, false, false));
returnValue.Append(separatorToUse);
```

File: src/System.Management.Automation/engine/MshObject.cs [hunk 4]
```diff
@@ -1144,8 +1150,15 @@ internal static string ToStringEnumerable(ExecutionContext context, IEnumerable
...
-                    returnValue.Append(PSObject.ToString(context, mshObj, separator, format, formatProvider, false, false));
+                    if (TryFastTrackPrimitiveTypes(obj, format, formatProvider, out string objString))
+                    {
+                        returnValue.Append(objString);
+                    }
+                    else
+                    {
+                        PSObject mshObj = PSObject.AsPSObject(obj);
+                        returnValue.Append(PSObject.ToString(context, mshObj, separator, format, formatProvider, false, false));
+                    }
```
Comment (daxian-dbw) [inline src/System.Management.Automation/engine/MshObject.cs]: Same here.

**Review by daxian-dbw — CHANGES_REQUESTED** [review]

**Comment by daxian-dbw** [comment]
Tests should be added to verify that both `ToStringEnumerator` and `ToStringEnumerable` honor the custom definition of `ToString` in ETS except for primitive types and the `string` type.

File: src/System.Management.Automation/engine/MshObject.cs (hunk 3 above)
Comment (iSazonov) [reply src/System.Management.Automation/engine/MshObject.cs]: 1. What about Decimal?
2. The PR intention is performance. In your sample we still do some duplication type checks. I'd refactor TryFastTrackPrimitiveTypes() in my sample with using IsPrimitive()

File: src/System.Management.Automation/engine/MshObject.cs [hunk 5]
```diff
@@ -1122,18 +1122,24 @@ internal static string ToStringEnumerator(ExecutionContext context, IEnumerator
             while (enumerator.MoveNext())
```
Comment (PoshChan) [inline src/System.Management.Automation/engine/MshObject.cs]: @iSazonov, your last commit had 1 failures in `PowerShell-CI-windows`
Enter-PSHostProcess tests.By Process Id.Can enter using NamedPipeConnectionInfo
```powershell
Exception calling "Invoke" with "0" argument(s): "The runspace state is not valid for this operation."
at <ScriptBlock>, D:\a\1\s\test\powershell\Modules\Microsoft.PowerShell.Core\Enter-PSHostProcess.Tests.ps1: line 136
136:                 $ps.AddScript('$pid').Invoke() | Should -Be $pwshId

```

**Comment by iSazonov** [comment]
I did not find how make a test for ToStringEnumerator().

File: src/System.Management.Automation/engine/MshObject.cs (hunk 3 above)
Comment (daxian-dbw) [reply src/System.Management.Automation/engine/MshObject.cs]: In my mind, a perf change shouldn't result in any behavioral changes, and that's why I'm concerned about you being aggressive here.
I would much prefer to be conservative
- Only treat string specially and keep on wrapping other values into `PSObject` **if we go for the most conservative manner**;
- Only treat `IsPrimitive` objects specially and keep wrapping other values **if we go a bit less conservative**

As for `decimal`, it's OK that we don't cover it. Again, it's not a functionality change that we should cover all scenarios, but a perf change that targets improve common scenarios.
As for `IntPtr` being a primitive type, I think it's fine too, as it's not really possible to use `IntPtr` directly in PowerShell script anyways.

My personal preference: only treat string specially. /cc @JamesWTruher @rjmholt what's your thoughts on this?

Comment (iSazonov) [reply src/System.Management.Automation/engine/MshObject.cs]: >  a perf change shouldn't result in any behavioral changes, and that's why I'm concerned about you being aggressive here

I am full agree and very appreciate your help!
The "aggressive" come from original issue where we consider primitive types too. I agree that for join operator strings are the most expected types for the operator. Nevertheless, I believe that there are many scenarios (like custom output formatting) when users will use the join operator in loops with primitive types. Notice that we change only an intermediate code and we added tests to exclude breaking change. So I believe we could be less conservative here. But I do not found how test the enumerator change and I'd revert the change.

**Comment by JamesWTruher** [comment]
The thing I'm most concerned about is the breaking nature of this change (and not being labeled initially as such). I think that we need to be very careful about addressing performance issues which result in behavioral change. And in this case, I'm not sure it's needed. I hope that we can query for type extensions (and specifically a `ToString`), without wrapping something in a psobject. If there's no extension for the type, then I totally agree we shouldn't wrap. But if there is, we need to use it.

I think there's far too much utility in ETS to turn our back on them. It doesn't matter if I can't see why anyone would do that, I'm sure our customers would be able to provide a number of reasons. That's the nature of a platform - it will be used in ways we can't imagine.

**Comment by iSazonov** [comment]
@JamesWTruher Thanks! I see your point about ETS and a breaking change.
It looks very general for the PR - we only bypass ETS  for primitive types and strings. The repo docs classifiers this as "unlikely breaking change". (Notice that there already is TryFastTrackPrimitiveTypes() for some code paths.)

Please see the follow example:
```powershell
Add-Type -TypeDefinition @"
namespace TestNS
{
   public class TestClass
   {
       public override string ToString()
       {
           return "1";
       }
   }
}
"@

$a1=[TestNS.TestClass]::new()

$a1
# Output is "1"

$a1 -join ";"
# Output is "1"

Update-TypeData -TypeName "TestNS.TestClass" -MemberType ScriptMethod -MemberName "ToString" -Value { return "test"}

$a1
# Output is "test"

$a1 -join ";"
# Output is "1" - unexpected?

$a1 -like "1"
# Output is "True" - unexpected?

$a1,$a1 -like "1"
# Output is - unexpected?
# "1"
# "1"
```
The example demonstrates two inconsistency (possible fix will be a breaking change):
- `-join` with one argument works another - it calls PSObject.ToStringParser() without wrapping to psobject and ignores ETS
- `-like` (and other string operators) also uses PSObject.ToStringParser() and ignores ETS

Now question is - should we have a consistency here and accept a breaking change, 
in other words, must `-join` work as other operators and ignore ETS too? Or do we should review and fix all string operators (-join, -like, -match,...) to honor ETS?

File: src/System.Management.Automation/engine/MshObject.cs (hunk 3 above)
Comment (rjmholt) [reply src/System.Management.Automation/engine/MshObject.cs]: > /cc @JamesWTruher @rjmholt what's your thoughts on this?

I think specialising for `string` only makes the most sense. It's by far the most common, and the place where we don't break things. I don't think there are many people out there doing lots of string joins on other primitive types, but by breaking ETS we'd lose functionality

**Comment by iSazonov** [comment]
@daxian-dbw It seems we have gathered the opinions of all who wanted to speak out. Common conclusion is a concern about ETS and that only string type could be fixed. Based on my previous comment I thing the string type case is also subject to the concern and I think close the PR and related issue as "By-design" with workaround to use `[string]::join()`.
